sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import java.util.UUID
+
+import org.apache.spark.SparkException
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.expressions.PredicateHelper
+import org.apache.spark.sql.catalyst.plans.logical.{AppendData, Command, LogicalPlan, OverwriteByExpression, OverwritePartitionsDynamic}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.catalog.Table
+import org.apache.spark.sql.connector.write.{LogicalWriteInfoImpl, SupportsDynamicOverwrite, SupportsOverwrite, SupportsTruncate, Write, WriteBuilder}
+import org.apache.spark.sql.execution.datasources.DataSourceStrategy
+import org.apache.spark.sql.sources.{AlwaysTrue, Filter}
+
+/**
+ * A rule that constructs [[Write]]s.
+ */
+object V2Writes extends Rule[LogicalPlan] with PredicateHelper {
+
+  import DataSourceV2Implicits._
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
+    case AppendData(relation: DataSourceV2Relation, query, options, _) =>
+      val writeBuilder = newWriteBuilder(relation.table, query, options)
+      val write = writeBuilder.build()
+      V2BatchWriteCommand(write, query)
+
+    case OverwriteByExpression(relation: DataSourceV2Relation, deleteExpr, query, options, _) =>
+      // fail if any filter cannot be converted. correctness depends on removing all matching data.
+      val filters = splitConjunctivePredicates(deleteExpr).map {
+        filter => DataSourceStrategy.translateFilter(deleteExpr,
+          supportNestedPredicatePushdown = true).getOrElse(
+          throw new AnalysisException(s"Cannot translate expression to source filter: $filter"))
+      }.toArray
```

Comment:
By the change, we move catalyst expression -> sources.Filter conversion to logical plans. So we will see both catalyst expressions and sources.expressions in logical plans in optimization.

Is it more clear if we use only catalyst expressions in logical plans, and convert to sources.Filter in physical plans when we need to interact datasources?

Reply:
And later in V2WriteRequirements, we also need to convert sources.Filter back to catalyst expressions.

Reply:
I don't think that it is possible to do the conversion later, because we need the write builder to be fully configured to produce a `Write` in this rule. That way, the `Write` can expose its order and distribution requirements in the other rule.

I think it is fine to convert to `Filter` here. That's the public API for filter expressions, so I don't think there is a requirement for it to be used only in physical plans. Filter is already used in the optimizer on the read path as well, because the `Scan` is similarly built and added to an optimizer plan so that the optimizer can handle stats based on the pushed filters.